### DISCUSSION TITLE
how to use loop in test runner

### DISCUSSION THREAD

**Discussion by DELEPaps** [webdriverio/webdriverio#8802] [Q&A]
I switched from 'Run in a script' to 'Test runner' because I wanted to use the 'Page Object Pattern'.

how do i implement this in 'Test runner'

I create an instance loop and each instance opens a different profile with a predefined range of values.

```
(async () => {
  let browser: any;

  let laps = 10;
  let profiles = 5;

  let countLaps = 1;
  let countProfile = 0;

  do {
    try {
      browser = await remote({
        capabilities: {
          browserName: "chrome",
          "goog:chromeOptions": {
            args: [
              `user-data-dir=/path/to/dir`,
              `profile-directory=Profile ${
                          countProfile === profiles - 0
                            ? (countProfile = 0)
                            : countProfile++
              }`,
              "start-maximized",
            ],
          },
        },
        waitforTimeout: 60000,
        connectionRetryTimeout: 120000,
        connectionRetryCount: 3,
      });
      await browser.url("chrome://version");
      await comment(browser);
      await browser.deleteSession();
    } catch (error) {
     await browser.deleteSession();
    }
    ++countLaps;
  } while (countLaps <= laps);
})();
```

**Comment by portyu9** [comment]
Hi @DELEPaps 

You don't need to create `remote()` sessions in a loop when using the WDIO test runner.

The runner is designed to own the browser sessions. Define the Chrome profiles as capabilities and put the actual test in a normal `it()` block.

For repeating the specs, current WebdriverIO also has a runner option:

```bash
npx wdio run ./wdio.conf.js --repeat 10
```

If you need different Chrome profiles, configure them per capability with `goog:chromeOptions`.

One important warning: Chrome user-data directories should not be shared by multiple simultaneously running Chrome processes. If these profiles run in parallel, give each worker its own `--user-data-dir`. Otherwise run the profile capabilities sequentially.

Also, the Page Object Pattern does **not** require the WDIO test runner — page objects are just JavaScript/TypeScript classes — although the runner is usually the more convenient structure for a test suite.

So I would remove the `remote()` / `deleteSession()` loop and model the repetitions and browser configurations at runner/configuration level instead.

I reviewed the complete script in the discussion; there were no replies, screenshots, or attachments.

Please mark this answer as accepted if it helped, thank you.
